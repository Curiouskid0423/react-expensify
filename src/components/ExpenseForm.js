import React from "react";
import moment from "moment";
import 'react-dates/initialize';
import { SingleDatePicker } from "react-dates";

/**
 * ExpenseForm will be reused in Edit Expense and Add Expense.
 * `react-dates` has some outdated dependencies. Beware.
 * @author Kevin Li
 */

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        const isEdit = !!props.expense;
        this.state = {
            description: isEdit ? props.expense.description : "",
            amount: isEdit ?
                (props.expense.amount / 100).toString() : "",
            note: isEdit ? props.expense.note : "",
            createdAt: isEdit ? moment(props.expense.createdAt) : moment(),
            calFocused: false,
            error: ""
        }
    }

    // Event listener for description change
    onDescChange = (e) => {
        const description = e.target.value;
        this.setState({ description });
    }

    // Event listener for amount changes, with regex check
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{1,2})?$/)) {
            this.setState({ amount });
        }
    }

    // Event listener for note changes
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState({ note });
    }

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState({ createdAt });
        }
    }

    onAddExpense = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            const error = "Please fill out the form.";
            this.setState({ error });
        } else {
            this.setState({error: ""});
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount),
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    }

    render() {
        return (
            <div>
                { this.state.error && <h4>{ this.state.error }</h4> }
                <form onSubmit = {this.onAddExpense}>
                    <input type = "text" placeholder = "Description"
                           value = {this.state.description} autoFocus = {true}
                           onChange={this.onDescChange}
                    />
                    <input type = "number" placeholder = "Amount" value={this.state.amount}
                           onChange={this.onAmountChange}/>
                    <br />
                    <textarea placeholder = "Optional notes for the expense"
                              value = {this.state.note}
                              onChange={this.onNoteChange}>
                    </textarea>
                    <SingleDatePicker date = {this.state.createdAt}
                                      onDateChange = {this.onDateChange}
                                      focused = {this.state.calFocused}
                                      onFocusChange = {({ focused }) => this.setState({calFocused: focused})}
                                      numberOfMonths = {1} isOutsideRange = {() => false}
                    />
                    <br/>
                    <button type="submit"> Submit </button>
                </form>
            </div>
        )
    }
}