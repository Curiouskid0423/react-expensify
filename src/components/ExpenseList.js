import React from "react";
import { connect } from "react-redux";
import ExpenseItem from "./ExpenseListItem";
import { getVisibleExpenses } from "../selectors/expenses";

/**
 * Helper method to load each expense.
 * @param props
 * @return {*}
 */
const ExpensesLoader = (props) => {
    if (props.expenses.length === 0) {
        return (<h4>No expense is tracked under the given condition.</h4>);
    }
    return props.expenses.map((item) => <ExpenseItem key = {item.id} {...item} />);
}

export const ExpenseList = (props) => {
    return (
        <div>
            <ExpensesLoader expenses = {props.expenses} />
        </div>
    )
}

/**
 * MapStateToProps take in two arguments: state, and ownProps
 * The `state` object could actually be called store (redux)
 * @param state
 * @return a `props` object
 */
const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);