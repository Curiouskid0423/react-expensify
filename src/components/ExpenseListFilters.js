import React from "react";
import {connect} from "react-redux";
import {setEndDate, setStartDate, setTextFilter, sortByAmount, sortByDate} from "../actions/filters";
import { DateRangePicker } from "react-dates";

class ExpenseListFilters extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            calFocused: null
        }
    }

    /**
     * Helper method to update sortBy with select options.
     * @param props is the `props` passed in by ExpenseListFilters
     * @param value
     */
    updateSortBy = (props, value) => {
        if (value === "amount") {
            this.props.dispatch(sortByAmount());
        } else {
            this.props.dispatch(sortByDate());
        }
    }

    onDatesChange = ({startDate, endDate}) => {
        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch(setEndDate(endDate));
    }

    onFocusChange = (calFocused) => {
        this.setState(() => ({ calFocused }));
    }

    /**
     * Dummy helper method to generate date id for react-dates library
     * @param moment
     * @return an id for the passed in moment object
     */
    getDateId = (moment) => {
        return moment ? moment.toString() : "emptyDate";
    }

    render() {
        return (
            <div>
                <input type = "text" onChange={(e) => {
                    this.props.dispatch(setTextFilter(e.target.value));
                }}/> &nbsp;
                <select onChange = {(e) => this.updateSortBy(this.props, e.target.value)}>
                    <option value = "date"> Date </option>
                    <option value = "amount"> Amount </option>
                </select>
                <DateRangePicker
                    startDate = { this.props.filters.startDate }
                    startDateId = {this.getDateId(this.props.filters.startDate)}
                    endDate = { this.props.filters.endDate }
                    endDateId = {this.getDateId(this.props.filters.endDate)}
                    onDatesChange = { this.onDatesChange }
                    focusedInput = {this.state.calFocused}
                    onFocusChange = {this.onFocusChange}
                    isOutsideRange = {() => false}
                    numberOfMonths = {1} showClearDates = {true}
                />
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

export default connect(mapStateToProps)(ExpenseListFilters);