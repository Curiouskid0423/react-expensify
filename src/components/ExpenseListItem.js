import React from "react";
import {Link} from "react-router-dom";


const ExpenseItem = ({ id, description,
                         amount, createdAt }) => {
    return (
        <div>
            <Link to = {`/edit/${id}`}>
                <h4>{ description }</h4>
            </Link>
            <h4> {amount} -- {createdAt} </h4>
        </div>
    );
};

export default ExpenseItem;