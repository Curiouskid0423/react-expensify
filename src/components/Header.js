import React from "react";

/**
 * All functional components also have to be Capitalized to be considered a
 * React component, whereas a class-based component needs to be Caps and extending
 * React.Component to be considered.
 */

const Header = (props) => {
    return (
        <div className = "header">
            <h1 className = "header__title">{props.title}</h1>
            <h2 className = "header__subtitle">{props.subtitle}</h2>
        </div>
    );
};

Header.defaultProps = {
    title: "Hello I am Kevin"
}

export default Header;