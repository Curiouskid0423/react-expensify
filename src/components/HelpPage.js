import React from "react";

const HelpPage = () => {
    return (
        <div>
            <h4> Help page. </h4>
        </div>
    )
};

export default HelpPage;