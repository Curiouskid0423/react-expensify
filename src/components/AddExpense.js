import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import {addExpense} from "../actions/expenses";

const AddExpense = (props) => {
    return (
        <div>
            <br/>
            <ExpenseForm onSubmit = {(expenseObj) => {
                props.dispatch(addExpense(expenseObj));
                props.history.push("/");
            }}/>
        </div>
    )
};

// TODO: We can safely use `connect` without `mapStateToProps` if we don't need
//  info from the store other than the function `dispatch()`.
export default connect()(AddExpense);