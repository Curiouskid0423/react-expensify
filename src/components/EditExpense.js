import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import {startEditExpense, startRemoveExpense} from "../actions/expenses";
import Button from "@material-ui/core/Button";

export class EditExpense extends React.Component {
    constructor(props) {
        super(props);
    }

    onEdit = (expense) => {
        const id = this.props.expense.id;
        this.props.dispatchEdit(id, expense);
        this.props.history.push("/");
    }

    onRemoval = () => {
        this.props.dispatchRemove({ id: this.props.expense.id});
        this.props.history.push("/");
    }

    render() {
        return (
            <div>
                <div className={"page-header"}>
                    <div className={"content-container"}>
                        <h1 className={"page-header__title"}>Edit Expense</h1>
                    </div>
                </div>
                    <ExpenseForm expense = {this.props.expense} onSubmit = {this.onEdit} />
                    {/*<button onClick = {this.onRemoval}> Remove </button>*/}
                    <div className={"remove-btn"}>
                        <Button variant="contained" type ={"submit"} onClick={this.onRemoval}>
                            Remove Expense
                        </Button>
                    </div>
                <br />
            </div>
        )
    }

}

// match.params.id is implicitly returned by React Component.
// ownProps take in the props from the connected component.
const mapStateToProps = (state, ownProps) => {
    return {
        expense: state.expenses.find((item) => {
            return item.id === ownProps.match.params.id;
        })
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        dispatchEdit: (id, expense) => dispatch(startEditExpense(id, expense)),
        dispatchRemove: (id) => dispatch(startRemoveExpense(id))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);