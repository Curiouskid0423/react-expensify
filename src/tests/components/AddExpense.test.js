import React from "react";
import { shallow } from "enzyme";
import { AddExpense } from "../../components/AddExpense";

describe("Add expense page defaults.", () => {
   it("should create a default snapshot.", () => {
       const wrapper = shallow(<AddExpense />);
       expect(wrapper).toMatchSnapshot();
   });
});