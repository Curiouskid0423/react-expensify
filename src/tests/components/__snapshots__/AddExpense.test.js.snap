// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add expense page defaults. should create a default snapshot. 1`] = `
<div>
  <br />
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
